transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacketLogger.java

```diff
@@ -0,0 +1,218 @@
+import java.util.Locale;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+
+/**
+ * A {@link ChannelLogger} that logs packets sent or received via a {@link TunTapChannel}.
+ */
+public class TunTapPacketLogger extends LoggingHandler {
+
+    private boolean _logPacketData = true;
+
+    public TunTapPacketLogger() {
+        super(LogLevel.INFO);
+    }
+
+    public TunTapPacketLogger(LogLevel level) {
+        super(level);
+    }
+
+    public TunTapPacketLogger(String name) {
+        super(name);
+    }
+
+    public TunTapPacketLogger(String name, LogLevel level) {
+        super(name, level);
+    }
+
+    public boolean logPacketData() {
+        return _logPacketData;
+    }
+
+    public void setLogPacketData(boolean logPacketData) {
+        _logPacketData = logPacketData();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET RECEIVED");
+            ctx.fireChannelRead(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET SENT");
+            ctx.write(msg, promise);
+        } else {
+            super.write(ctx, msg, promise);
+        }
+    }
+
+    private void logPacket(TunTapPacket packet, boolean isEtherFrame, String eventName) {
+        if (logger.isEnabled(internalLevel)) {
+            logger.log(internalLevel, formatPacket(packet, isEtherFrame, eventName, _logPacketData));
+        }
+    }
+
+    public static String formatPacket(TunTapPacket packet, boolean isEtherFrame, String title, boolean logPacketData) {
+        StringBuilder out = new StringBuilder();
+        Formatter formatter = new Formatter(out, Locale.US);
+        ByteBuf packetData = packet.packetData();
+        int protocol = packet.protocol();
+        int ipPacketOffset = 0;
+
+        formatter.format("%s (protocol %s, len %d):\n",
+                title,
+                TunTapPacket.getProtocolName(packet.protocol()),
+                packetData.readableBytes());
+
+        // If the packet is an Ethernet frame...
+        if (isEtherFrame) {
+
+            // Generally a sender doesn't specify a protocol when sending an Ethernet frame, since
+            // Linux doesn't require it.  Therefore extract the protocol from the Ethernet frame data.
+            protocol = ((int) packetData.getUnsignedByte(packetData.readerIndex() + 12)) << 8 |
```

Comment:
It's the same: `protocol = packetData.getUnsignedShort(packetData.readerIndex() + 12);` ?